### DISCUSSION TITLE
Prisma Edge Vercel, Prisma, VercelDB, PrismaClientUnknownRequestError

### DISCUSSION THREAD

**Discussion by sjashwin** [prisma/prisma#22281] [Q&A]
### Bug description

```
Fetching Quizzes PrismaClientUnknownRequestError: 
Invalid `prisma.quiz.findMany()` invocation:


Error occurred during query execution:
ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(PostgresError { code: "26000", message: "prepared statement \"s412\" does not exist", severity: "ERROR", detail: None, column: None, hint: None }), transient: false })
    at Ei.handleRequestError (/Users/ashwin/Documents/GitHub/Scholar-Sprint/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/edge.js:84:7381)
    at Ei.handleAndLogRequestError (/Users/ashwin/Documents/GitHub/Scholar-Sprint/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/edge.js:84:6595)
    at Ei.request (/Users/ashwin/Documents/GitHub/Scholar-Sprint/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/edge.js:84:6293)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async u (/Users/ashwin/Documents/GitHub/Scholar-Sprint/node_modules/.pnpm/@prisma+client@5.6.0_prisma@5.6.0/node_modules/@prisma/client/runtime/edge.js:89:10383)
    at async fetchQuizzes (webpack-internal:///(rsc)/./app/actions/quiz.ts:59:20)
    at async Dashboard (webpack-internal:///(rsc)/./app/dashboard/page.tsx:47:21) {
  clientVersion: '5.6.0'
  ```

### How to reproduce

Prisma Accelerate 

datasource db {
  provider          = "postgresql"
  url               = env("POSTGRES_PRISMA_ACCELERATE")
  directUrl         = env("POSTGRES_URL_NON_POOLING") // uses a direct connection
  shadowDatabaseUrl = env("POSTGRES_URL_NON_POOLING") // used for migrations
}

### Expected behavior

Should fetch data for the given query. Should work consistently. But, this issue occurs from time to time.

### Prisma information

```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```

### Environment & setup

- OS: MacOS
- Database: VercelDB
- Node.js version: v20.9.0

### Prisma Version

```
'5.6.0'
```

**Comment by janpio** [comment]
> ConnectorError(ConnectorError { user_facing_error: None, kind: QueryError(PostgresError { code: "26000", message: "prepared statement \"s412\" does not exist", severity: "ERROR", detail: None, column: None, hint: None }), transient: false })

You are using Prisma Accelerate, so you use the `@prisma/client/edge` import and have a `prisma://` url configured as `POSTGRES_PRISMA_ACCELERATE`, correct?
If so, which connection URL from Vercel did you put into the connection string in Prisma Data Platform? The pooled or the unpooled one? If you used the pooled one, did you add `pgbouncer=true` as another query string?

Aside:

> shadowDatabaseUrl = env("POSTGRES_URL_NON_POOLING") // used for migrations

Where do you have this snippet from? This will lead to data loss. Please let us know where this is still documented so we can remove it.

**Reply by sjashwin** [reply]
I'm using prisma edge client
This is how I've created the above singleton.
```
import { PrismaClient } from '@prisma/client/edge';
import { withAccelerate } from '@prisma/extension-accelerate';

const prismaClientSingleton = () => {
  return new PrismaClient().$extends(withAccelerate());
};

type PrismaClientSingleton = ReturnType<typeof prismaClientSingleton>;

const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClientSingleton | undefined;
};

const prisma = globalForPrisma.prisma ?? prismaClientSingleton();

export default prisma;

if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma;
```

```
shadowDatabaseUrl 
```

Isn't documented anywhere, recently. Forgot to remove it. Thanks for reminding.


Yes, I'm using non-pooling URL. The accelerate url is prefixed with `prisma://`.

The issue occurs periodically and is not persistent (does not always occur). Sometime the this error occurs, otherwise, it is all fine. I didn't have this issue last week.

**Reply by sjashwin** [reply]
This is how my api endpoint is written for the above error query:

```ts
import prisma from '@/lib/prisma';
import { Prisma } from '@prisma/client/edge';

export const runtime = 'edge';

const quizWithAuthor = Prisma.validator<Prisma.QuizDefaultArgs>()({
  include: {
    author: true
  }
});

export type QuizWithAuthor = Prisma.QuizGetPayload<typeof quizWithAuthor>;

export async function GET(request: Request) {
  const { id } = await request.json();
  let quiz: QuizWithAuthor[] | null;
  if (id === undefined) {
    quiz = await prisma.quiz.findMany({
      include: {
        author: true
      },
      cacheStrategy: {
        ttl: 60 * 60
      }
    });
  } else {
    quiz = await prisma.quiz.findMany({
      where: {
        authorId: id
      },
      include: {
        author: true
      }
    });
  }
  if (!quiz) return { status: 404, body: { message: 'Not found' } };
  return { status: 200, body: quiz };
}
```

**Reply by sjashwin** [reply]
App Router and NextJS 14.0.3

**Reply by janpio** [reply]
The error message `prepared statement \"s412\" does not exist"` indicates that there is PgBouncer active, but `pgbouncer=true` is not being used. Please make sure to update the connection string in Accelerate again, either with a pooled one with `pgbouncer=true` added or really using the unpooled one (preferred, as Accelerate does its own pooling already).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
